test/extensions/filters/http/dynamic_forward_proxy/proxy_filter_integration_test.cc

```diff
@@ -151,6 +156,46 @@ name: envoy.clusters.dynamic_forward_proxy
   bool use_cache_file_{};
 };
 
+int64_t getHeaderValue(const Http::ResponseHeaderMap& headers, absl::string_view name) {
+  EXPECT_FALSE(headers.get(Http::LowerCaseString(name)).empty());
+  int64_t val;
+  if (!headers.get(Http::LowerCaseString(name)).empty() &&
+      absl::SimpleAtoi(headers.get(Http::LowerCaseString(name))[0]->value().getStringView(),
+                       &val)) {
+    return val;
+  }
+  return 0;
+}
+
+void ProxyFilterIntegrationTest::testConnectionTiming(IntegrationStreamDecoderPtr& response,
+                                                      bool cached_dns, int64_t original_usec) {
+  int64_t dns_start = getHeaderValue(response->headers(), "dns_start");
+  int64_t dns_end = getHeaderValue(response->headers(), "dns_end");
+  int64_t connect_start = getHeaderValue(response->headers(), "upstream_connect_start");
+  int64_t connect_end = getHeaderValue(response->headers(), "upstream_connect_complete");
+  int64_t handshake_end = getHeaderValue(response->headers(), "upstream_handshake_complete");
+  int64_t request_send_end = getHeaderValue(response->headers(), "request_send_end");
+  int64_t response_begin = getHeaderValue(response->headers(), "response_begin");
+  Event::DispatcherImpl dispatcher("foo", *api_, timeSystem());
+
+  auto time = timeSystem().monotonicTime();
+  int64_t now = time.time_since_epoch().count();
+
+  ASSERT_LT(original_usec, dns_start);
+  ASSERT_LE(dns_start, dns_end);
+  if (cached_dns) {
+    ASSERT_GE(dns_end, connect_start);
+  } else {
+    ASSERT_LE(dns_end, connect_start);
+  }
+  ASSERT_LE(connect_start, connect_end);
+  ASSERT_LE(connect_end, handshake_end);
+  ASSERT_LE(handshake_end, request_send_end);
+  ASSERT_LE(request_send_end, response_begin);
+  ASSERT_LT(handshake_end, timeSystem().monotonicTime().time_since_epoch().count());
+  ASSERT_EQ(now, time.time_since_epoch().count());
```

Comment:
Is this guaranteed to be true since some time will have elapsed between line 182 and here?

Reply:
oh sorry, that was me testing my understanding of monotonic time when I was debugging out of order times.  Because it's latched, it shouldn't change as time passes.  But now that I've verified I understand it I can remove that.